presto-test-jdbc-compatibility-old-server/src/test/java/io/prestosql/TestJdbcResultSetCompatibility.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logging;
+import io.prestosql.jdbc.BaseTestJdbcResultSet;
+import io.prestosql.jdbc.PrestoDriver;
+import org.testcontainers.containers.PrestoContainer;
+import org.testng.SkipException;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Factory;
+import org.testng.annotations.Test;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+@Test(singleThreaded = true)
+public class TestJdbcResultSetCompatibility
+        extends BaseTestJdbcResultSet
+{
+    private static final int NUMBER_OF_TESTED_VERSIONS = 5;
+    private static final int TESTED_VERSIONS_GRANULARITY = 3;
+
+    private final Optional<String> testedPrestoVersion;
+    private PrestoContainer<?> prestoContainer;
+
+    @Factory(dataProvider = "testedPrestoVersions")
+    public TestJdbcResultSetCompatibility(Optional<String> testedPrestoVersion)
+    {
+        this.testedPrestoVersion = requireNonNull(testedPrestoVersion, "testedPrestoVersion is null");
+    }
+
+    @DataProvider
+    public static Object[][] testedPrestoVersions()
+    {
+        try {
+            String currentVersionString = PrestoDriver.class.getPackage().getImplementationVersion();
+            if (currentVersionString == null) {
+                return new Object[][] {
+                        {Optional.empty()}
```

Comment:
Current version is tested by `io.prestosql.jdbc.TestJdbcResultSet`.
I understand you need this to run the test from intellij (to test the test),
so please add an appropriate comment. Otherwise it looks like "this test
past versions... or it doesn't"

Also, how do we actually verify the test works?
Would it work if we allowed `currentVersionString == null` only when some
additional sys property is set? You would set it manually when running from IDE
and CI or when running from maven would use the JDBC jar and fall into
`currentVersionString != null` category.

Also, maybe we can simplify this by baking in the current version via resources?
Like here
https://github.com/prestosql/presto/blob/cd061b3410c752253d64f096176bf226a3dea11a/presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/Launcher.java#L170-L172

Reply:
I need currentVersion to determine the list of older versions to be tested.
If current version cannot be determined there will be only single incarnation of test class with `testedPrestoVersion` set to `empty()`. 
Then initialization of test class will fail via:
```
if (testedPrestoVersion.isEmpty()) {
            throw new AssertionError("Could not determine current Presto version");
        }
```

This is needed so we notice on CI if for some reason we cannot determine a list of versions to be tested. We do not a green run there.

As for testing in IntelliJ it is not supported right now. Do we care very much? I can do do similar mechanism we have in product-tests-launcher but I am not sure if that is worth it. Maybe it would be enough to allow overriding currentPrestoVersion via environment variable and explain how to do that in massage of assertion error we throw when version cannot be determined?

Reply:
> As for testing in IntelliJ it is not supported right now. Do we care very much?

Yes, otherwise, debugging and fixing breakages becomes much harder.